File: dynamice programming/198.py
# Google, Amazon, AirBnB, OYO, Microsoft
# Stickler Thief
# Why DP?
# We have options to choose
# Find Optimal (max/min)
# Overlapping Sub-problems

# Recursion + Memoization (Top-Down)
# Tc, Sc = O(n)
class Solution:
    def __init__(self):
        self.t = [-1] * 101

    def solve(self, nums, i, n):
        if i >= n:
            return 0
        if self.t[i] != -1:
            return self.t[i]
        steal = nums[i] + self.solve(nums, i + 2, n)
        skip = self.solve(nums, i + 1, n)
        self.t[i] = max(steal, skip)
        return self.t[i]

    def rob(self, nums):
        n = len(nums)
        self.t = [-1] * 101
        return self.solve(nums, 0, n)
    
# Bottom-Up Approach
# Define-> t[i] = max stolen money till house i
class Solution:
    def rob(self, nums):
        n = len(nums)
        if n == 1:
            return nums[0]
        
        # Initialize the memoization array
        t = [0] * (n + 1)
        
        # Base cases
        t[0] = 0  # No house, no money
        t[1] = nums[0]  # Only one house
        
        for i in range(2, n + 1):
            steal = nums[i - 1] + t[i - 2]  # Money if we rob the current house
            skip = t[i - 1]  # Money if we skip the current house
            t[i] = max(steal, skip)  # Max money we can have up to house i
        
        return t[n]  # Return the maximum money that can be robbed

File: dynamice programming/198.cpp
// Top-down
class Solution {
public:
    int t[101];
    int solve(vector<int>& nums, int i, int n){
        if(i >= n)
            return 0;
        if(t[i] != -1)
            return t[i];
        int steal = nums[i] + solve(nums, i+2, n);
        int skip = solve(nums, i+1, n);
        return t[i] = max(steal, skip);
    }
    int rob(vector<int>& nums) {
       int n = nums.size();
       memset(t, -1, sizeof(t));
       return solve(nums, 0, n);
    }
};

// Bottom-Up
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n==1) 
            return nums[0];
        vector<int> t(n+1, 0);
        // t[i] = max stolen money till house i
        // no house, no steal: i = 0
        t[0] = 0;
        // 1 house: i = 1
        t[1] = nums[0];
        for(int i = 2; i <= n; i++){
            int steal = nums[i-1] + t[i-2]; 
            int skip = t[i-1];
            t[i] = max(steal, skip);
        }
        return t[n];
    }
};